proposals/0023-asynchronous-program-execution-and-broadcast.md

```diff
@@ -0,0 +1,280 @@
+vote transactions
+
+* VoteHash - like BankHash, but for vote only transactions.
+
+* BundleTransactions - a transaction signed by the builder that can
+list transactions in the builders UserBlock to be executed in
+one ordered batch. This transaction can also add a priority fee
+such that the entire bundle is prioritized for execution as a batch.
+This transaction can only appear in the builders UserBlock and can
+only reference transactions prior to it in the UserBlock. TBD on
+format.
+
+## Detailed Design
+
+### Overview
+Leaders are scheduled to build blocks as they are currently by
+the LeaderSchedule.
+
+Builder's are scheduled along side leaders to build UserBlocks -
+blocks of non vote transactions. N number of builders can be scheudled
+concurrently to build blocks.
+
+While a leader is scheduled, they receive and encode votes as normal.
+Any well formed UserBlocks that were received from the previous or
+current UserBlockSlot are added to the leaders PoH ledger as
+UserBlockEntry.
+
+The N concurrent Builder create blocks out of user transactions.
+These are transmitted to the cluster via turbine.  The leader
+receives and decodes them and generates a UserBlockEntry, and adds
+it to PoH as soon as the leaders PoH has started the UserBlockSlot.
+
+Validators can vote on leader blocks by executing the votes, but
+before completing the execution of UserBlock transactions.
+
+### Fork Choice
+
+If a validator doesn't have the builder's UserBlock, the validator
+doesn't vote on the proposed block and tries to repair the UserBlock.
+That fork is ignored for fork choice.
+
+Otherwise the validator evaluates the forks as usual. 
+
+### UserBlock schedule
+
+There are 2 UserBlockSlots per slot. Builders produce 1 UserBlock
+per UserBlockSlot. So if the cluster is configured with 3 Builders,
+each leader slot would see a total of 6 UserBlocks per slot.  3 in
+the first half of the leaders slot, and 3 in the second half.
+
+The current leader schedule rotates each leader every 4 slots, but
+builders could be rotated at a different schedule then the leaders
+and rotation could be staggered.
+
+### UserBlock Compute Limits
+
+If the overall compute capacity for user transactions per leader
+block is 48m CU, and cluster is configured with 2 builders, then
+each UserBlock can use no more then 48m/4 or 12m CU.
+
+### UserBlock execution
+
+Execution of UserBlocks can happen asynchronously to voting.  When
+voting validators transmit their most recent BankHash, but it may
+be for an older parent slot then the slot they are voting on.
+
+Each UserBlock is assumed to have been created simultaneously during
+the UserBlockSlot that it was encoded by the leader.
+
+For each UserBlock in the PoH section spanning the UesrBlockSlot,
+the transactions are ordered by priority fee before execution.
+
+If two transactions from two different blocks have the same priority,
+they are ordered by which UserBlock appears first in the leaders
+PoH.
```

Comment:
Does this effectively creates something similar to the Ethereum mempool, but on chain?
As people are going to publish their transactions, yet the final order is determined only at the end of the `UserBlockSlot`.
Someone observing transactions sent to Builders can reliably front run any of those transactions, by including a transaction with a higher fee, as long as the current UserBlockSlot is not over yet.
And they can do so using any of the available Builders.
I am not necessarily criticizing - more like trying to make sure my understanding of the proposed mechanics is correct.

Or is the assumption that Builders do not share the transaction information until the end of the UserBlockSlot?

Reply:
200ms is very short, I would assume that they just need to stream transactions as they arrive